User transformation test

rspriyamdixit opened the pull request:
1. Integration test cases added for User Transformation and Other server endpoints(track, screen, page, screen, etc).
2. Validation added under webhook and postgres destination.
## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

lvrach commented on docker_test.go:
```diff
@@ -41,11 +50,65 @@ import (
+	writeKey    string
+}
+
+type WareHouseDestinationTest struct {
```
```suggestion
type WarehouseDestinationTest struct {
```

lvrach commented on docker_test.go:
```diff
@@ -194,10 +263,51 @@ func blockOnHold() {
+		fmt.Println(err)
+		return "", err
+	}
+	req.Header.Add("Authorization", "Basic cnVkZGVyOnBhc3N3b3Jk")
```
what can someone do with this key ? Are we ok exposing this ?

lvrach commented on docker_test.go:
```diff
@@ -194,10 +263,51 @@ func blockOnHold() {
+	log.Println("Sending pixel/v1/page Event")
+	url := fmt.Sprintf("http://localhost:%s/pixel/v1/page?writeKey=%s&anonymousId=identified_user_id", httpPort, writeKey)
+	method := "GET"
+	resBody, _ := GetEvent(url, method)
```
error not handled, consider returning err or passing t *testing.T to this method so you can do t.Fail()

lvrach commented on docker_test.go:
```diff
@@ -194,10 +263,51 @@ func blockOnHold() {
+	url := fmt.Sprintf("http://localhost:%s/pixel/v1/page?writeKey=%s&anonymousId=identified_user_id", httpPort, writeKey)
+	method := "GET"
+	resBody, _ := GetEvent(url, method)
+	log.Println(resBody)
+	log.Println("pixel/v1/page Event Sent Successfully")
```
```suggestion
```

Debug log, should it remain in the test suite ?

lvrach commented on docker_test.go:
```diff
@@ -194,10 +263,51 @@ func blockOnHold() {
+	url = fmt.Sprintf("http://localhost:%s/pixel/v1/track?writeKey=%s&anonymousId=identified_user_id&event=product_reviewed_again", httpPort, writeKey)
+	method = "GET"
+	resBody, _ = GetEvent(url, method)
+	log.Println(resBody)
+	log.Println("pixel/v1/track Event Sent Successfully")
```
error handling and excessive logging

lvrach commented on docker_test.go:
```diff
@@ -417,6 +535,16 @@ func run(m *testing.M) (int, error) {
+
+	initWHConfig()
+
+	defer SetWHPostgresDestination(pool)()
```
I don't like the fact that we are using and setting globals, what if ` SetWHPostgresDestinatio` was to be `SetupWHPostgresDestinatio` and return a struct (also this struct could have a method for cleanup).

rspriyamdixit replied:
@achettyiitr

lvrach commented on docker_test.go:
```diff
@@ -425,7 +553,7 @@ func run(m *testing.M) (int, error) {
 		ExposedPorts: []string{"9090"},
 		Env: []string{
-			"CONFIG_BACKEND_URL=https://api.dev.rudderlabs.com",
+			"CONFIG_BACKEND_URL=https://api.rudderlabs.com",
```
Do we really want to do this in production ?

rspriyamdixit replied:
Just keeping it for time being, will discuss with Jayc to how to reduce this dependency.

lvrach commented on docker_test.go:
```diff
@@ -473,6 +601,11 @@ func run(m *testing.M) (int, error) {
 	webhookurl = webhook.Server.URL
 	log.Println("webhookurl", webhookurl)
 
+	webhook2 = NewWebhook()
```
webhook2 can be confusing, what if webhookWithUT ?

lvrach commented on docker_test.go:
```diff
@@ -473,6 +601,11 @@ func run(m *testing.M) (int, error) {
+	webhook2 = NewWebhook()
+	defer webhook2.Close()
+	webhookurl2 = webhook2.Server.URL
+	log.Println("webhookurl2", webhookurl2)
```
```suggestion
```

lvrach commented on docker_test.go:
```diff
@@ -660,12 +806,118 @@ func TestWebhook(t *testing.T) {
+			}
+		]
+	}`)
+	SendEvent(payload_Batch, "batch", writeKey)
```
Should send event accept a reader or a string, if most of the cases we are using `strings.NewReader` before calling it

rspriyamdixit replied:
Not sure if this is possible as we need JSON payload and call type both.
Will think of approach where we can pick the JSON payload for external file in testdata folder on the bases of call type.

lvrach commented on docker_test.go:
```diff
@@ -660,12 +806,118 @@ func TestWebhook(t *testing.T) {
+		return len(webhook.Requests()) == 10
+	}, time.Minute, 30*time.Millisecond)
+
+	i := -1
+	require.Eventually(t, func() bool {
+		i = i + 1
+		req := webhook.Requests()[i]
+		body, _ := io.ReadAll(req.Body)
+		return gjson.GetBytes(body, "anonymousId").Str == "anonymousId_1"
```
I am not sure of this logic here, doesn't seem safe to me

rspriyamdixit replied:
Tested the approach, works well.
But yes ideally it's not the optimal code.
Will think of some other implementation and include in upcoming PR.

lvrach commented on docker_test.go:
```diff
@@ -746,9 +1037,11 @@ out:
-			require.Contains(t, string(msg.Value), "identified user id")
+			require.Equal(t, "identified_user_id", string(msg.Key))
+			require.Contains(t, string(msg.Value), "identified_user_id")
+			// TODO: Verify User Transformation
+			
+			// TODO: Verify Destination Transformation
```
TODOs

lvrach commented on docker_test.go:
```diff
@@ -795,4 +1088,1001 @@ func consume(topics []string, master sarama.Consumer) (chan *sarama.ConsumerMess
+}
+
+// whGatewayTest Checking for gateway jobs
+func whGatewayTest(t *testing.T, wdt *WareHouseDestinationTest) {
```
In the following tests, can you add some additional context as what we expect that should happen ?

rspriyamdixit replied:
Sure, will ask Akash to add some context.

lvrach commented on docker_test.go:
```diff
@@ -795,4 +1088,1001 @@ func consume(topics []string, master sarama.Consumer) (chan *sarama.ConsumerMess
+	require.Contains(t, wdt.whEventsCountMap, "gateway")
+	gwEvents := wdt.whEventsCountMap["gateway"]
+
+	// Checking for the gateway jobs
```
Instead of comments in the code you can do this, in case this test fails the printed logs could help with debugging.
```suggestion
	t.Log("Checking for the gateway jobs")
```

lvrach reviewed (changes requested):
Please, improve PR description. Explaining your testing approach and goals. Lots of line of code, hard to review correct without proper context.

rspriyamdixit commented:
> Please, improve PR description. Explaining your testing approach and goals. Lots of line of code, hard to review correct without proper context.

Thanks Leo for detailed review.
 Addressed most of the comments, will take care of rest in upcoming PR.

lvrach commented on main.go:
```diff
@@ -338,8 +338,9 @@ func Run(ctx context.Context) {
 		stats.StopRuntimeStats()
-
-		os.Exit(1)
+		if config.GetEnv("RUDDER_GRACEFUL_SHUTDOWN", "true") == "true" {
```
Not to confuse this with enabling or disable graceful timeout:
```suggestion
		if config.GetEnv("RUDDER_GRACEFUL_SHUTDOWN_TIMEOUT_EXIT", "true") == "true" {
```

lvrach commented on main.go:
```diff
@@ -339,8 +339,9 @@ func Run(ctx context.Context) {
 		stats.StopRuntimeStats()
-
-		os.Exit(1)
+		if config.GetEnv("RUDDER_GRACEFUL_SHUTDOWN_TIMEOUT_EXIT", "true") == "true" {
```
```suggestion
		if config.GetBool("RUDDER_GRACEFUL_SHUTDOWN_TIMEOUT_EXIT") == true {
```

kuldeep0020 commented on docker_test.go:
```diff
@@ -185,9 +188,15 @@ func randString(n int) string {
-	anonymous_id string
-	user_id      string
-	count        string
+	anonymous_id		 string
```
Need gofmt here

kuldeep0020 commented on docker_test.go:
```diff
@@ -185,9 +188,15 @@ func randString(n int) string {
+	anonymous_id		 string
+	user_id    			 string
+	count      		 	 string
+	context_myuniqueid	 string
```
here also

kuldeep0020 commented on docker_test.go:
```diff
@@ -185,9 +188,15 @@ func randString(n int) string {
+	context_ip 			 string
+	prop_key			 string
+	myuniqueid	 		 string
+	ip 					 string
```
here also
